sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcStreamingWriteSupport.scala

```diff
@@ -0,0 +1,238 @@
+import org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils.makeSetter
+import org.apache.spark.sql.jdbc.JdbcDialects
+import org.apache.spark.sql.sources.v2.writer.{DataWriter, WriterCommitMessage}
+import org.apache.spark.sql.sources.v2.writer.streaming.{StreamingDataWriterFactory, StreamingWriteSupport}
+import org.apache.spark.sql.types.StructType
+
+/**
+ * Dummy commit message. The DataSourceV2 framework requires a commit message implementation but we
+ * don't need to really send one.
+ */
+case object JdbcWriterCommitMessage extends WriterCommitMessage
+
+/**
+ * A [[StreamingWriteSupport]] for jdbc writing.
+ * Responsible for generating the writer factory.
+ */
+class JdbcStreamingWriteSupport(
+    schema: StructType,
+    options: Map[String, String]
+  ) extends StreamingWriteSupport with Logging {
+
+  override def commit(epochId: Long, messages: Array[WriterCommitMessage]): Unit = {
+    logInfo(s"epoch ${epochId} of JdbcStreamWriter commited!")
+  }
+  override def abort(epochId: Long, messages: Array[WriterCommitMessage]): Unit = {
+    logInfo(s"epoch ${epochId} of JdbcStreamWriter aborted!")
+  }
+  override def createStreamingWriterFactory(): StreamingDataWriterFactory = {
+    new JdbcStreamWriterFactory(schema, options)
+  }
+}
+
+/**
+ * A [[StreamingDataWriterFactory]] for jdbc writing.
+ * Will be serialized and sent to executors to generate the per-task data writers.
+ */
+case class JdbcStreamWriterFactory(
+    schema: StructType,
+    options: Map[String, String]
+  ) extends StreamingDataWriterFactory {
+
+  override def createWriter(
+    partitionId: Int,
+    taskId: Long,
+    epochId: Long): DataWriter[InternalRow] = {
+    JdbcStreamDataWriter(schema, options)
+  }
+}
+
+/**
+ * A [[DataWriter]] for Jdbc writing.
+ * One data writer will be created in each partition to process incoming rows.
+ */
+case class JdbcStreamDataWriter(
+    schema: StructType,
+    options: Map[String, String]
+  ) extends DataWriter[InternalRow] with Logging {
+
+  private val jdbcOptions = new JDBCOptions(options)
+
+  // use a local cache for batch write to jdbc.
+  private val batchSize = jdbcOptions.batchSize
+  private val localBuffer = new ArrayBuffer[Row](batchSize)
+  private val maxRetryNum = jdbcOptions.maxRetryNumber
+
+  // the first part is the column name list, the second part is the placeholder string.
+  private val sqlPart: (String, String) = {
+    val columnListBuilder = new StringBuilder()
+    val holderListBuilder = new StringBuilder()
+    schema.fields.foreach { field =>
+      columnListBuilder.append(",").append(field.name)
+      holderListBuilder.append(",?")
+    }
+    assert(columnListBuilder.nonEmpty, "Empty schema of data!")
+    (columnListBuilder.substring(1), holderListBuilder.substring(1))
+  }
+
+  private val sql: String = s"REPLACE INTO ${jdbcOptions.tableOrQuery}" +
+    s" ( ${sqlPart._1} ) values ( ${sqlPart._2} )"
+
+  logInfo(s"Sql string for jdbc writing is ${sql}")
```

Comment:
Is this useful for users? If this is only for developers, let's not use `logInfo`.

Reply:
OK, I will change to logTrace